### PR TITLE
Add pr-shepherd-mcp MCP server with webhook push notifications

### PR DESCRIPTION
**Pull request by jonathanong**
## Summary

- Adds a new `pr-shepherd-mcp` binary — a long-lived MCP server that exposes all pr-shepherd CLI commands as 10 MCP tools, enabling Claude Code to call them directly without spawning subprocesses
- Adds an optional HTTP webhook endpoint (`SHEPHERD_WEBHOOK_PORT`, default 3000) that receives GitHub PR events and forwards them to the connected session as MCP `notifications/message` wrapped in `<github-webhook-activity>` tags — the same format Claude Code's PR activity event system already handles
- Adds `shepherd_subscribe_pr` / `shepherd_unsubscribe_pr` tools so the agent can control which PRs trigger notifications in the current session

## New files

| File | Purpose |
|------|---------|
| `src/mcp-index.mts` | Entry point; reads `SHEPHERD_WEBHOOK_PORT` + `GITHUB_WEBHOOK_SECRET` env vars |
| `src/mcp/server.mts` | MCP server setup, tool registration, webhook→notification bridge |
| `src/mcp/subscriptions.mts` | In-memory `Set<prNumber>` — which PRs forward events to this session |
| `src/mcp/webhook.mts` | HTTP server; validates `X-Hub-Signature-256`; extracts PR number from payload |
| `src/mcp/tools.mts` | Dispatcher + `buildToolList()` combining both tool-list files |
| `src/mcp/tool-list-pr.mts` | Tool definitions: check, resolve_fetch, resolve_mutate, commit_suggestion, iterate |
| `src/mcp/tool-list-ops.mts` | Tool definitions: monitor, status, log_file, subscribe_pr, unsubscribe_pr |
| `src/mcp/tool-handlers.mts` | Handler implementations delegating to existing `run*()` functions |
| `src/mcp/tool-coerce.mts` | Input coercion helpers (optNum, reqStr, optStringArray, etc.) |
| `docs/mcp-server.md` | Full setup + tool reference documentation |

## Modified files

- `package.json` — adds `@modelcontextprotocol/sdk ^1.29.0` dep; adds `pr-shepherd-mcp` bin entry
- `scripts/build.mjs` — writes `bin/pr-shepherd-mcp` shebang wrapper + chmod + node_modules/.bin symlink
- `README.md` — adds "MCP Server" section with quick-start config snippet

## Design notes

- **Separate binary**: `pr-shepherd-mcp` is a clean entry point that doesn't call `setupLog()` (which would interfere with the MCP stdio transport by patching `process.stdout`)
- **Tool output**: all tools return `--format=text` output as MCP `TextContent`, identical to what the CLI emits
- **Webhook disabled by default opt-out**: set `SHEPHERD_WEBHOOK_PORT=0` to skip the HTTP server entirely
- **Existing CLI unchanged**: no shared mutable state between the CLI and MCP paths; all 802 existing tests still pass

## Test plan

- [x] `npm run build` produces `bin/pr-shepherd-mcp`
- [x] MCP handshake + ListTools returns all 10 tools
- [x] `SHEPHERD_WEBHOOK_PORT=0` starts without HTTP server
- [x] `npm run lint` / `npm run typecheck` / `npm run format:check` all pass
- [x] `npm test` — 802 tests pass, no regressions

https://claude.ai/code/session_016zr21ZuUx6kjymkK2uQjat

---
_Generated by [Claude Code](https://claude.ai/code/session_016zr21ZuUx6kjymkK2uQjat)_

## Shepherd Journal

**Iteration 1 (2026-04-27) — addressed review feedback from @copilot-pull-request-reviewer and @jonathanong**

Applied:
- [PRRT_kwDOSGizTs5965-9](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149346942) (`reqNumArray` empty-after-filter): changed both `reqNumArray` and `optStringArray` to throw on wrong-type elements rather than filtering them out.
- [PRRT_kwDOSGizTs5965_U](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149346982) (docs notification type): updated `docs/mcp-server.md` to say "MCP logging notification" and corrected the "Per-connection → Per-process" scope claim.
- [PRRT_kwDOSGizTs5965_l](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149347006) (`shepherd_resolve_mutate` no-op): added explicit guard in `handleResolveMutate` — returns MCP error when all three ID lists are empty; also guards `dismissReviewIds` without `dismissMessage`.
- [PRRT_kwDOSGizTs5965_s](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149347021) (path filter): webhook now rejects non-`/` / non-`/webhook` paths with 404.
- [PRRT_kwDOSGizTs5965_4](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149347042) (body size limit): added 5 MB cap; responds 413 and destroys the request if exceeded.
- [PRRT_kwDOSGizTs5966AC](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149347060) (`onEvent` try/catch): wrapped `buildPayload` + `onEvent` in try/catch; logs to stderr on error, always responds 204.
- [PRRT_kwDOSGizTs597DK8](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149397881) (webhook fails open): added loud stderr WARNING at startup when `GITHUB_WEBHOOK_SECRET` is unset; changed missing/invalid signature status to 401.
- [PRRT_kwDOSGizTs597DXt](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149398515) (SIGINT/SIGTERM): handlers now `await webhookServer.stop()` and call `process.exit(0)`.
- [PRRT_kwDOSGizTs597Do2](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149399665) (bind failure): `start()` now catches EADDRINUSE, logs to stderr, and continues with webhooks disabled instead of crashing the entire MCP server.
- [PRRT_kwDOSGizTs597D1Y](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149401161) (subscription scope docs): fixed "Per-connection" → "Per-process" in the docs table.
- [PRRT_kwDOSGizTs597D5m](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149401473) (`repoFullName` sentinel): made `repoFullName` optional on `WebhookPayload`; omit the `repo:` line when absent.
- [PRRT_kwDOSGizTs597ECM](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149402565) / [PRRT_kwDOSGizTs597EGD](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149402897): subscribe/unsubscribe responses now echo only the affected PR number, not the full subscription list.

Deferred (design-level, not addressed in this iteration):
- [PRRT_kwDOSGizTs597DSF](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149398515) (`notifications/message` undocumented): this is a Claude Code host behavior. The docs have been updated to remove the circular "same format handled by…" claim. Whether `level: "info"` logging notifications reach the session context is a host implementation detail — needs verification against Claude Code internals.
- [PRRT_kwDOSGizTs597DgC](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149399861) (`buildPayload` strips context): surfacing the raw event body is a larger design change. Current 6-field summary keeps notification size minimal; a follow-up could add event-type-specific fields (e.g., review comment body for `pull_request_review_comment` events).
- [PRRT_kwDOSGizTs597Dul](https://github.com/jonathanong/pr-shepherd/pull/145#discussion_r3149401161) (subscription per-connection clearing): `StdioServerTransport` doesn't expose an `onclose` hook in the current SDK version. Deferred until the SDK exposes transport lifecycle events.

### CODE REVIEW

File: src/mcp/webhook.mts [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+      req.on("end", () => {
+        const body = Buffer.concat(chunks);
+
+        if (this.opts.secret) {
+          const sig = req.headers["x-hub-signature-256"];
+          if (typeof sig !== "string" || !verifySignature(body, sig, this.opts.secret)) {
+            res.writeHead(400).end("Invalid signature");
+            return;
+          }
+        }
```
Comment (jonathanong) [inline src/mcp/webhook.mts:78]: When `GITHUB_WEBHOOK_SECRET` is unset, this endpoint accepts any anonymous POST and forwards arbitrary `<github-webhook-activity>` blocks into the connected Claude Code session — that's a prompt-injection vector for anyone who can reach port 3000.

Two suggestions, in order of effort:
1. Write a loud `pr-shepherd-mcp: WARNING — GITHUB_WEBHOOK_SECRET unset; webhook auth disabled` to stderr at server startup so the user notices before exposing the port.
2. When no secret is configured, bind to `127.0.0.1` only (or refuse to start the HTTP server entirely and require an explicit opt-in env var to enable unauthenticated mode).

Also: GitHub's webhook validation docs recommend rejecting a missing `X-Hub-Signature-256` header with HTTP 401/403, not 400 — the current `"Invalid signature"` message on a missing header may confuse debugging.

File: src/mcp/server.mts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      secret: opts.webhookSecret,
+      onEvent: (payload) => {
+        if (!subscriptions.isSubscribed(payload.prNumber)) return;
+        server
+          .sendLoggingMessage({ level: "info", data: formatWebhookNotification(payload) })
+          .catch((e: unknown) => {
+            process.stderr.write(
+              `pr-shepherd-mcp: failed to send webhook notification: ${String(e)}\n`,
+            );
+          });
```
Comment (jonathanong) [inline src/mcp/server.mts:72]: I cross-checked Claude Code's public MCP docs (`https://code.claude.com/docs/en/mcp`). The docs describe `list_changed` notifications, but there's no documented behavior for `notifications/message` (logging notifications) being surfaced into the agent context — and no documented `<github-webhook-activity>` tag format.

This means the push-notification feature works only because of an undocumented Claude Code host behavior. If the host filters logging notifications by level, or changes how it handles arbitrary `data` strings, this path silently regresses.

Two requests:
1. Either link to a public spec documenting the `<github-webhook-activity>` contract, or soften the phrase "the same format handled by Claude Code's PR activity event system" in `docs/mcp-server.md` — right now it reads as a circular hand-wave with no external reference.
2. Confirm `level: "info"` actually reaches the session. Some MCP hosts filter by level; if uncertain, raise to `"notice"` or `"warning"` and document the verified behavior.

File: src/mcp/server.mts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+    await webhookServer.start();
+    process.stderr.write(`pr-shepherd-mcp: webhook server listening on port ${opts.webhookPort}\n`);
+
+    process.on("SIGTERM", () => {
+      void webhookServer.stop();
+    });
+    process.on("SIGINT", () => {
+      void webhookServer.stop();
+    });
```
Comment (jonathanong) [inline src/mcp/server.mts]: These handlers fire-and-forget `webhookServer.stop()` and never:
- `await` the stop promise (process may exit before `server.close()` finishes draining sockets),
- close the MCP `Server` / `StdioServerTransport`,
- call `process.exit(0)`.

Suggest:
```ts
process.on("SIGTERM", () => { void (async () => { await webhookServer.stop(); process.exit(0); })(); });
process.on("SIGINT",  () => { void (async () => { await webhookServer.stop(); process.exit(0); })(); });
```
Also worth handling a double-SIGINT with a synchronous `process.exit(1)` fallback so a user who Ctrl-C's twice isn't left with a stuck process.

File: src/mcp/webhook.mts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+
+        res.writeHead(204).end();
+      });
+      req.on("error", () => res.writeHead(400).end("Request error"));
+    });
+
+    await new Promise<void>((resolve, reject) => {
+      this.server!.listen(this.opts.port, () => resolve());
+      this.server!.once("error", reject);
+    });
```
Comment (jonathanong) [inline src/mcp/webhook.mts]: If port 3000 is already in use (common when running multiple worktrees or after an unclean shutdown), `listen()` emits `error`, `start()` rejects, and `mcp-index.mts` catches the rejection and calls `process.exit(1)` — killing the entire MCP server, including the tool dispatcher that doesn't need the port.

Recommend treating bind failure as a soft warning instead: catch `EADDRINUSE` here, log to stderr (`pr-shepherd-mcp: webhook port N in use, notifications disabled`), set `this.server = null`, and return normally. The user gets a degraded experience (no push notifications) rather than a dead server.

File: src/mcp/webhook.mts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  prNumber: number,
+): WebhookPayload {
+  const repo = body["repository"] as { full_name?: string } | undefined;
+  const sender = body["sender"] as { login?: string } | undefined;
+  const headCommit = body["head_commit"] as { id?: string } | undefined;
+
+  const payload: WebhookPayload = {
+    event,
+    prNumber,
+    repoFullName: repo?.full_name ?? "unknown/unknown",
```
Comment (jonathanong) [inline src/mcp/webhook.mts]: The hardcoded sentinel `"unknown/unknown"` will surface in the agent's context as `repo: unknown/unknown` for events without a `repository.full_name` (pings, replays, malformed payloads). The agent may try to interpret it literally.

CLAUDE.md "Lean output" says to omit fields when they carry no actionable value. `action`, `ref`, `sha`, and `actor` are all conditionally appended in `formatWebhookNotification` — make `repoFullName` optional on `WebhookPayload` too, and skip the `repo:` line when absent, consistent with the other optional fields.

File: src/mcp/server.mts [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+
+export async function startMcpServer(opts: McpServerOptions): Promise<void> {
+  const version = readVersion();
+
+  const server = new Server(
+    { name: "pr-shepherd", version },
+    { capabilities: { tools: {}, logging: {} } },
+  );
+
+  const subscriptions = new SubscriptionStore();
```
Comment (jonathanong) [inline src/mcp/server.mts:39]: `SubscriptionStore` is created once per *process* and is never cleared. The `docs/mcp-server.md` "Differences from CLI" table describes subscription state as "Per-connection in-memory set," but that's not what the code does — it's per-process.

If the stdio transport disconnects and reconnects within the same process (e.g., Claude Code restart without killing the MCP server), old subscriptions persist across sessions. Two options:
- **(a) Make it actually per-connection:** hook `transport.onclose = () => subscriptions.clear()` after `server.connect(transport)`.
- **(b) Accept per-process and update the doc:** change the table entry to "Per-process in-memory set" so the contract matches reality.

File: src/mcp/webhook.mts [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+  const ref = body["ref"];
+  if (typeof ref === "string") payload.ref = ref;
+
+  const sha = body["after"] ?? headCommit?.id;
+  if (typeof sha === "string") payload.sha = sha;
+
+  if (sender?.login) payload.actor = sender.login;
+
+  return payload;
+}
```
Comment (jonathanong) [inline src/mcp/webhook.mts:193]: CLAUDE.md says "Surface data, don't classify it … prefer to ship the raw fields and any context the agent would otherwise have to fetch separately."

`buildPayload` cherry-picks 6 fields from a typically rich GitHub event payload, silently dropping review comment text, check suite conclusion URLs, workflow run names, sender profile URLs, etc. The agent receiving a `pull_request_review_comment` notification still has to call `shepherd_resolve_fetch` to see what the comment actually said — the webhook already had it.

Suggest including a verbatim excerpt or the full raw body (perhaps as a JSON string field `raw` capped at a reasonable byte limit) so the agent can avoid the round-trip for the common cases. At minimum, for `pull_request_review` and `pull_request_review_comment` events, surface the comment body.

File: src/mcp/tool-coerce.mts [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+  return v.filter((x): x is string => typeof x === "string");
+}
+
+export function reqNumArray(input: Record<string, unknown>, key: string): number[] {
+  const v = input[key];
+  if (!Array.isArray(v) || v.length === 0) {
+    throw new Error(`${key} must be a non-empty array of numbers`);
+  }
+  return v.filter((x): x is number => typeof x === "number");
+}
```
Comment (jonathanong) [inline src/mcp/tool-coerce.mts:51]: If an agent passes `prNumbers: ["123"]` (string instead of number — a common LLM coercion mistake), the non-empty array check on line 40 passes, but `.filter` on line 43 returns `[]`. The downstream call then runs as a silent no-op with no error surfaced.

Recommend throwing on any wrong-type element rather than filtering them out:
```ts
return v.map((x, i) => {
  if (typeof x !== "number") throw new Error(`${key}[${i}] must be a number, got ${typeof x}`);
  return x;
});
```
Same issue applies to `optStringArray` on line 35.

File: src/mcp/tool-list-pr.mts [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+          description: "Review node IDs to dismiss",
+        },
+        dismissMessage: { type: "string", description: "Required when dismissing reviews" },
+        requireSha: {
+          type: "string",
+          description: "Wait for GitHub to receive this commit SHA before resolving",
+        },
+      },
+    },
+  },
```
Comment (jonathanong) [inline src/mcp/tool-list-pr.mts:71]: The description says "Provide at least one of `resolveThreadIds`, `minimizeCommentIds`, or `dismissReviewIds`," and line 64 says `dismissMessage` is "Required when dismissing reviews" — but neither constraint is enforced by the schema or by `handleResolveMutate`.

- An empty-body call silently no-ops (no error surfaced).
- A `dismissReviewIds` call without `dismissMessage` likely fails deep in the GraphQL layer with a less-than-helpful error rather than an early validation error.

Two approaches:
1. Add a JSON Schema `anyOf` (or `minProperties: 1`) for the at-least-one constraint and let the SDK validate.
2. Add explicit guards at the top of `handleResolveMutate`: throw if all three arrays are empty/missing, throw if `dismissReviewIds` is non-empty and `dismissMessage` is absent.

File: src/mcp/tool-handlers.mts [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+export function handleUnsubscribePr(
+  input: Record<string, unknown>,
+  subs: SubscriptionStore,
+): CallToolResult {
+  const prNumber = reqNum(input, "prNumber");
+  subs.unsubscribe(prNumber);
+  const all = subs.listSubscribed();
+  return ok(
+    `Unsubscribed from PR #${prNumber}.\nCurrently subscribed PRs: ${all.length > 0 ? all.join(", ") : "(none)"}`,
+  );
```
Comment (jonathanong) [inline src/mcp/tool-handlers.mts]: Every subscribe/unsubscribe response appends the full subscription list. For a long-lived session with many subscriptions the response grows unbounded, and there's no upper bound on `SubscriptionStore`'s underlying `Set` — repeated `shepherd_subscribe_pr` calls with arbitrary PR numbers are a trivial DoS on the agent's context.

Suggest echoing only the changed PR (e.g. `"Subscribed to PR #123."`) and dropping the full list from the response, or capping it at ~20 with a `(+N more)` suffix. The agent can always call `shepherd_status` to see current state.

**Review by jonathanong — COMMENTED** [review]
Reviewed against Claude Code MCP docs (https://code.claude.com/docs/en/mcp) and GitHub webhook validation docs (https://docs.github.com/en/webhooks/using-webhooks/validating-webhook-deliveries). Inline comments cover the material concerns; this body summarizes the cross-cutting themes.

## Headline concerns

1. **Webhook trust model (inline on `webhook.mts:51-57`).** When `GITHUB_WEBHOOK_SECRET` is unset the HTTP listener accepts any anonymous POST and injects the result into the agent context — a prompt-injection footgun for anyone who can reach port 3000. At minimum warn loudly at startup; ideally bind to loopback when no secret is set.

2. **Undocumented MCP notification surface (inline on `server.mts:60-66`).** Claude Code's public MCP docs only document `list_changed` notifications. There's no documented support for `notifications/message` being surfaced to the agent, no documented `<github-webhook-activity>` tag contract. The feature currently works because of an undocumented host behavior — worth either linking a spec or softening the docs claim.

3. **Lifecycle gaps (inlines on `server.mts:73-78` and `webhook.mts:81-84`).** SIGINT/SIGTERM fire-and-forget without awaiting or exiting; bind failure kills the entire MCP server instead of degrading gracefully.

4. **Surface data, don't classify it (inline on `webhook.mts:126-154`).** `buildPayload` strips the GitHub event to 6 fields, dropping review comment text and other context the agent would otherwise round-trip to fetch.

## Suggested follow-up tests

The PR already adds 7 test files with solid baseline coverage (HMAC valid/invalid/missing, port-0 disabled, all PR-extraction branches, dispatcher unknown-tool, all coercer happy paths). Edge cases worth adding:

1. **`webhook.test.mts` — short signature header.** Send `x-hub-signature-256: sha256=short` (different byte length than the 71-char expected value). Locks the `a.length !== b.length` early-return so a future refactor can't accidentally pass mismatched buffers to `timingSafeEqual` (which throws rather than returns false).
2. **`webhook.test.mts` — array-valued signature header.** Node coerces duplicate headers to `string[]`; verify the `typeof sig !== "string"` guard returns 400.
3. **`webhook.test.mts` — non-POST methods beyond GET.** Existing test only covers GET; add PUT/DELETE to pin the 405 contract.
4. **`webhook.test.mts` — `pull_request.number === 0` and `< 0`.** Locks the `prNumber > 0` guard so events with a zero/negative PR number are correctly dropped.
5. **`webhook.test.mts` — restart on same port.** `start() → stop() → start()` should not throw `EADDRINUSE` — validates that the lifecycle fix for #4 above doesn't break clean restart.
6. **`mcp-index.mock.test.mts` — `parsePort` edge cases.** `undefined → 3000`, `"abc" → 3000`, `"-1" → 3000`, `"3000.5" → 3000`. Currently exercised only through the integration path.
7. **`tool-coerce.test.mts` — `reqNumArray` mixed-type input.** `[1, "2", 3]` currently returns `[1, 3]` silently; if the throw-on-mismatch fix from inline #8 is adopted, add a test that pins the new error message.
8. **`tools.mock.test.mts` — non-`Error` throw path.** A handler throwing a plain string should still hit the `String(e)` branch in `buildToolHandler`; existing tests only cover `Error` instances.

## Out of scope / confirmed OK

- `readVersion()` path `../../package.json` relative to `import.meta.url`: resolves correctly both in source (`src/mcp/server.mts` → repo root) and after `tsc` build (`bin/mcp/server.mjs` → repo root). No issue.
- HMAC comparison on hex strings vs raw bytes: fine in practice since GitHub always sends a fixed-length 71-char `sha256=<hex>` header. Style nit only, not flagged inline.
- `shepherd_log_file` taking no inputs: consistent with the CLI subcommand.
